### PR TITLE
Add 'optimized_graph' to Function.

### PR DESCRIPTION
**Pull request by ZolotukhinM**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26703 Remove convert_to_ssa argument from runCleanupPasses - it is only used in one place.
* #26705 Use optimized_graph in graph_executor.
* #26489 Use optimized graph in Inline (essentially, making Inline recursive now).
* **#26488 Add 'optimized_graph' to Function.**
* #26571 Use std::mutex instead of std::call_once in Function when we initialize GraphExecutor.
* #26487 Remove 'recurse' parameter from Inline.

Currently the main use case for this graph is inlining and that's the
only optimization we perform. We probably should run more cleanups on
this graph in future.

Differential Revision: [D17485745](https://our.internmc.facebook.com/intern/diff/D17485745)

### CODE REVIEW

File: torch/csrc/jit/function.h
Comment (zdevito) [inline torch/csrc/jit/function.h]: Nothing is using this yet, so it would be confusing to commit it without hooking it up to the executors.

File: torch/csrc/jit/function.h
Comment (zdevito) [inline torch/csrc/jit/function.h]: This isn't thread safe. Functions are invokable from multiple threads (e.g. in Predictor).

**Review by zdevito — COMMENTED** [review]
This is the right direction, but needs to be hooked up more to flush out any potential gotchas. It would be confusing to add this and then not use it anywhere.

**Comment by zdevito** [comment]
I noticed now that there is more in this patch set. If they remain separate PRs they should at least land together so there isn't confusion about what they are being used for. We should also commit to using this in the executor before we land it. I want to avoid the situation where we end up with an additional optimization pathway that is used, but only for the inlined versions of functions.

**Comment by ZolotukhinM** [comment]
> I noticed now that there is more in this patch set. If they remain separate PRs they should at least land together so there isn't confusion about what they are being used for.

Yes, that's the plan, I'll wait until all PRs are accepted and then land them together.

> We should also commit to using this in the executor before we land it. I want to avoid the situation where we end up with an additional optimization pathway that is used, but only for the inlined versions of functions.

Agreed.

**Review by zdevito — APPROVED** [review]

**Pull request closed without merging**
